zkevm-circuits/src/evm_circuit/execution/extcodehash.rs

```diff
@@ -0,0 +1,330 @@
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        param::N_BYTES_ACCOUNT_ADDRESS,
+        step::ExecutionState,
+        table::{AccountFieldTag, CallContextFieldTag},
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{ConstraintBuilder, StepStateTransition, Transition::Delta},
+            from_bytes, Cell, RandomLinearCombination, Word,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    util::Expr,
+};
+use eth_types::ToLittleEndian;
+use eth_types::{evm_types::GasCost, Field, ToAddress, ToScalar, U256};
+use ethers_core::utils::keccak256;
+use halo2_proofs::{circuit::Region, plonk::Error};
+use lazy_static::lazy_static;
+
+lazy_static! {
+    static ref EMPTY_CODE_HASH_LE_BYTES: [u8; 32] = U256::from(keccak256(&[])).to_le_bytes();
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct ExtcodehashGadget<F> {
+    same_context: SameContextGadget<F>,
+    external_address: RandomLinearCombination<F, N_BYTES_ACCOUNT_ADDRESS>,
+    tx_id: Cell<F>,
+    is_warm: Cell<F>,
+    nonce: Cell<F>,
+    balance: Cell<F>,
+    code_hash: Cell<F>,
+    is_empty: Cell<F>, // boolean for if the external account is empty
+    nonempty_witness: Cell<F>, /* if the account isn't empty, will witness that by holding
+                        * inverse of one of balance, nonce, or RLC(code hash) -
+                        * RLC(EMPTY_CODE_HASH_LE_BYTES) */
+}
+
+impl<F: Field> ExecutionGadget<F> for ExtcodehashGadget<F> {
+    const NAME: &'static str = "EXTCODEHASH";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::EXTCODEHASH;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let external_address = cb.query_rlc();
+        cb.stack_pop(external_address.expr());
+
+        let tx_id = cb.call_context(None, CallContextFieldTag::TxId);
+
+        let is_warm = cb.query_bool();
+        cb.account_access_list_write(
+            tx_id.expr(),
+            from_bytes::expr(&external_address.cells),
+            1.expr(),
+            is_warm.expr(),
+            None,
```

Comment:
I've reviewed the implementation of SSTORE and I realized I missed this before.  For an op that can revert, this shouldn't be `None`.  Take a look here: 
https://github.com/appliedzkp/zkevm-circuits/blob/c5dda20e901951e61f751b89dd50607f83b98a95/zkevm-circuits/src/evm_circuit/execution/sstore.rs#L75-L81

This will require adding two read operations for `CallContextFieldTag::RwCounterEndOfReversion`, `CallContextFieldTag::IsPersistent` as seen here:
https://github.com/appliedzkp/zkevm-circuits/blob/c5dda20e901951e61f751b89dd50607f83b98a95/zkevm-circuits/src/evm_circuit/execution/sstore.rs#L50-L56

Reply:
I've made these changes to the circuit and the changes needed to the bus mapping, but now the circuit tests fail.

Do you know if the witness generation supports `RwCounterEndOfReversion` and `IsPersistent`? I noticed that the sload bus mapping doesn't correspond to the circuit and that the sload and sstore circuits both manually construct the rw tables for their tests.

Reply:
> I've made these changes to the circuit and the changes needed to the bus mapping, but now the circuit tests fail.
> 
> Do you know if the witness generation supports `RwCounterEndOfReversion` and `IsPersistent`? I noticed that the sload bus mapping doesn't correspond to the circuit and that the sload and sstore circuits both manually construct the rw tables for their tests.

I will investigate this

Reply:
Found the bug!  After you added generating the operations for CallContextFieldTag::RwCounterEndOfReversion and CallContextFieldTag::RwCounterEndOfReversion you altered the positions of the other operations for that step in rw_indices.  So when you query the values in the assign advice method, you need to update the indexes of rw_index to obtain the values for nonce, balance and code_hash. See https://github.com/appliedzkp/zkevm-circuits/pull/353#discussion_r832359566